src/java.net.http/share/classes/jdk/internal/net/http/quic/QuicCubicCongestionController.java

```diff
@@ -0,0 +1,170 @@
+        super(source, rttEstimator);
+        this.rttEstimator = rttEstimator;
+    }
+
+    @Override
+    public void packetSent(int packetBytes) {
+        lock.lock();
+        try {
+            super.packetSent(packetBytes);
+            if (isCwndLimited()) {
+                Deadline now = timeSource.instant();
+                if (lastFullWindow == null) {
+                    lastFullWindow = now;
+                } else {
+                    long timePassedNanos = Deadline.between(lastFullWindow, now).toNanos();
+                    if (timePassedNanos > 0) {
+                        /* "The elapsed time MUST NOT include periods during which cwnd
+                           has not been updated due to application-limited behavior"
+                           "A flow is application limited if it is currently sending less
+                            than what is allowed by the congestion window."
+
+                           We are sending asynchronously; one thread is sending data,
+                           a separate thread is processing the acknowledgements.
+                           We can't rely on cwnd being fully utilized when we process an ack, because
+                           most of the time it won't be.
+
+                           Instead, we assume that if we filled the cwnd, we were not application-limited
+                           in the last RTT (which is a pretty good approximation because of pacing),
+                           and acknowledgements for all packets sent prior to filling the cwnd
+                           count towards cwnd increase.
+                         */
+                        long rttNanos = TimeUnit.MICROSECONDS.toNanos(rttEstimator.state().smoothedRttMicros());
+                        timeNanos += Math.min(timePassedNanos, rttNanos);
+                        lastFullWindow = now;
+                    }
+                }
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+
+    protected boolean congestionAvoidanceAcked(int packetBytes, Deadline sentTime) {
+        boolean isAppLimited = sentTime.isAfter(lastFullWindow);
+        if (!isAppLimited) {
+            if (wEstBytes < cwndPriorBytes) {
+                wEstBytes += Math.max((long) (ALPHA * maxDatagramSize * packetBytes / congestionWindow), 1);
+            } else {
+                wEstBytes += Math.max((long)maxDatagramSize * packetBytes / congestionWindow, 1);
+            }
+            // target = Wcubic(t + RTT)
+            long rttNanos = TimeUnit.MICROSECONDS.toNanos(rttEstimator.state().smoothedRttMicros());
+            double dblTargetBytes = wCubicBytes(timeNanos + rttNanos);
+            long targetBytes;
+            // not sure if dblTarget can overflow a long, but 1.5 congestionWindow can not.
+            if (dblTargetBytes > 1.5 * congestionWindow) {
+                targetBytes = (long) (1.5 * congestionWindow);
```

Comment:
```suggestion
            // targetLimit is 1.5 * congestionWindow
            long targetLimit = congestionWindow + (congestionWindow >> 1)
            if (dblTargetBytes > targetLimit) {
                targetBytes = targetLimit;
```

Reply:
Refactored the code, I think it's easier to follow now.